File: 3sum.py
"""
Time Complexity : O(n^2)
Space Complexity : O(1)
Did this code successfully run on Leetcode :  yes 
Three line explanation of solution in plain english

// Your code here along with comments explaining your approach
"""


class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        target = 0
        result = []
        n = len(nums)
        nums.sort()
        for i in range(n):
            if i != 0 and nums[i] == nums[i - 1]:
                continue
            if nums[i] > target:
                break
            low = i + 1
            high = n - 1
            while low < high:
                sum = nums[i] + nums[low] + nums[high]
                if sum == target:
                    temp = [nums[i], nums[low], nums[high]]
                    result.append(temp)
                    low += 1
                    high -= 1
                    while low < high and nums[low] == nums[low - 1]:
                        low += 1
                    while low < high and nums[high] == nums[high + 1]:
                        high -= 1
                elif sum > target:
                    high -= 1
                else:
                    low += 1
        return result

File: 3_Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
    int n = nums.size();
    vector<vector<int>> results;
    sort(nums.begin(), nums.end());
    for( int i =0;i<n-2;i++){
        if(i>0 && nums[i]==nums[i-1])continue;
        int l = i+1;
        int r= n-1;
        while(l<r){
            int currSum = nums[i]+nums[l]+nums[r];
            if(currSum==0){
                results.push_back({nums[i], nums[l], nums[r]});
                l++;
                r--;
                while(l<r && nums[r]== nums[r+1]){
                r--;
            }
            while(l<r && nums[l]== nums[l-1]){
                l++;
            }
            }
            
            else if(currSum > 0){
                r--;
            }
            else{
                l++;
            }
        }

    }
    return results;   
    }
};




//sort(nums.begin(), nums.end());
        
        
    //     vector<vector<int>> result;
    //    // if(nums[0]>0)return {};
    //    // if(nums.size()<3)return{};
    //     for(int i=0;i<nums.size()-2;i++){
        
    //     int low=i+1;
    //     int high = nums.size()-1;
    //     int target = -nums[i];
    //     if (i > 0 && nums[i] == nums[i - 1]) {
    //             continue; // Skip duplicates
    //         }
    //     while(low<high){
    //         int sum = nums[low]+ nums[high];
    //         if(target>sum){
    //             low++;
    //         }else if(sum>target){
    //             high--;
    //         }
    //         else{
    //             result.push_back({nums[i], nums[low], nums[high]});
    //             while(low<high && nums[low]==nums[low+1]) low++;
    //             while(low<high && nums[high]==nums[high-1]) high--;
    //             low++;
    //             high--;
    //         }
    //     }
    //     }
    //    return result; 
